File: scripts/vcf-table.pl
#!/usr/bin/env perl

use strict;

my $HELP = <<HELP;

  Convert VCF file to tab-delimited table for merging with ANNOVAR output.

  usage: vcf-table.pl file.vcf sample_name > vcf_table.txt

HELP

if (!$ARGV[1]) {
	die $HELP;
}

main();

# main subroutine
sub main {
	my $vcf = $ARGV[0];
	my $sample = $ARGV[1];

	# get variant caller (sample can't be determined from some VCFs)
	# my ($caller, $sample) = get_info($vcf);
	my $caller = get_info($vcf);

	# print germline header
	if ($caller eq "HaplotypeCaller" || $caller eq "LoFreq") {
		print "#MUT\tSAMPLE\tCHR\tPOS\tQUAL\tDEPTH\tFREQ\n";
	}
	# print old MuTect somatic header
	if ($caller eq "MuTect114" || $caller eq "MuTect116") {
		print "#MUT\tSAMPLE N\tSAMPLE T\tCHR\tPOS\tt_lod_fstar ( log of ( likelihood event is real / likelihood event is seq error ) )\tN Depth\tN Freq\tT Depth\tT Freq\n";
	}
	# print somatic header (T and N are separate columns to make it clear which is which)
	if ($caller eq "MuTect2" || $caller eq "Strelka") {
		print "#MUT\tSAMPLE T\tSAMPLE N\tCHR\tPOS\tQUAL\tT DEPTH\tT FREQ\tN DEPTH\tN FREQ\n";
	}

	# remove header lines
	my @vcf_records = `cat $vcf | grep -v "^#" | LC_ALL=C sort -k1,1 -k2,2n`;
	while (my $line = shift(@vcf_records)) {
		chomp($line);
		my $clean_line;

		if ($caller eq "HaplotypeCaller") {
			$clean_line = format_haplotypecaller($line, $sample);
		}
		if ($caller eq "LoFreq") {
			$clean_line = format_lofreq($line, $sample);
		}
		if ($caller eq "MuTect114") {
			# $clean_line = format_mutect_114($line);
		}
		if ($caller eq "MuTect116") {
			# $clean_line = format_mutect_116($line);
		}
		if ($caller eq "MuTect2") {
			$clean_line = format_mutect2($line, $sample);
		}
		if ($caller eq "Strelka") {
			$clean_line = format_strelka($line, $sample);
		}


		# skip lines that returned empty
		if ($clean_line) {
			print $clean_line . "\n";
		}
	}

}

# get variant caller
sub get_info {
	my $vcf = $_[0];

	my $caller;

	open IN, $vcf or die $!;
	foreach my $line (<IN>) {
		chomp($line);

		# get caller
		if ($line =~ /ID=HaplotypeCaller/) {
			$caller = "HaplotypeCaller";
			last;
		}
		if ($line =~ /source=lofreq/) {
			$caller = "LoFreq";
			last;
		}
		if ($line =~ /muTector/) {
			# mutect actually uses the txt output instead of vcf (more data and better formatting)
			$caller = "MuTect114";
			last;
		}
		if ($line =~ /MuTect:1.1.6/) {
			# mutect actually uses the txt output instead of vcf (more data and better formatting)
			$caller = "MuTect116";
			last;
		}
		if ($line =~ /GATKCommandLine.MuTect2/) {
			$caller = "MuTect2";
			last;
		}
		if ($line =~ /source=strelka/) {
			$caller = "Strelka";
			last;
		}

	}
	close IN;

	return ($caller);
}

# format HaplotypeCaller line
sub format_haplotypecaller {
	my $line = $_[0];
	my $sample = $_[1];
	my $out;

	# split the needed columns (#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT [SAMPLE])
	my @cols = split(/\t/, $line);
	# sample fields (GT:AD:DP:GQ:PL)
	my @sample_fields = split(':', $cols[9]);
	# allelic depth field (Allelic depths for the ref and alt alleles in the order listed)
	my @ad_cols = split(',', $sample_fields[1]);

	# output columns
	my $chr = $cols[0];
	my $pos = $cols[1];
	my $ref = $cols[3];
	my $alt = $cols[4];
	my $depth = $sample_fields[2];

	# do not report if
	# there are not any quality reads (reported depth 0)
	# less than 5 variant call supporting reads
	if ( ($depth > 0) && ($ad_cols[1] >= 5) ) {
		($pos, $ref, $alt) = adjust_indels($pos, $ref, $alt);
		my $qual = (sprintf "%.1f", $cols[5]);
		my $freq = sprintf("%.3f", ( $ad_cols[1] / $depth ));
		$out = join("\t", "${chr}:${pos}:${ref}:${alt}", $sample, $chr, $pos, $qual, $depth, $freq);
	}

	$out;
}

# format lofreq line
sub format_lofreq {
	my $line = $_[0];
	my $sample = $_[1];
	my $out;

	# split the needed columns (#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT [SAMPLE])
	my @cols = split(/\t/, $line);

	# output columns
	my $chr = $cols[0];
	my $pos = $cols[1];
	my $ref = $cols[3];
	my $alt = $cols[4];
	my $qual = $cols[5];
	my $info = $cols[7];
	my $depth = $info;
	$depth =~ s/.*DP=(.+?);.*/$1/;
	my $freq = $info;
	$freq =~ s/.*AF=(.+?);.*/$1/;

	# do not report if frequency is less than 1%
	if ($freq > 0.01) {
		($pos, $ref, $alt) = adjust_indels($pos, $ref, $alt);
		$freq = sprintf("%.3f", $freq);
		$out = join("\t", "${chr}:${pos}:${ref}:${alt}", $sample, $chr, $pos, $qual, $depth, $freq);
	}

	$out;
}

# format MuTect 1.1.4 line
sub format_mutect_114 {
	my $line = $_[0];
	my $out;

	# split the needed columns (#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT [SAMPLE])
	my @cols = split(/\t/, $line);

	# output columns
	my $chr = $cols[0];
	my $pos = $cols[1];
	my $ref = $cols[3];
	my $alt = $cols[4];
	my $sample_t = $cols[5];
	my $sample_n = $cols[6];
	my $t_lod_fstar = sprintf("%.1f", $cols[16]);
	my $t_ref_count = $cols[20];
	my $t_alt_count = $cols[21];
	my $n_ref_count = $cols[30];
	my $n_alt_count = $cols[31];

	my $depth_t = $t_alt_count + $t_ref_count;
	my $freq_t = 0;
	if ($depth_t > 0) {
		$freq_t = sprintf("%.3f", ( $t_alt_count / $depth_t ));
	}

	my $depth_n = $n_alt_count + $n_ref_count;
	my $freq_n = 0;
	if ($depth_n > 0) {
		$freq_n = sprintf("%.3f", ( $n_alt_count / $depth_n ));
	}

	my $judgement = $cols[34];

	if ($judgement eq "KEEP") {
		$out = join("\t", "${chr}:${pos}:${ref}:${alt}", $sample_n, $sample_t, $chr, $pos, $t_lod_fstar, $depth_n, $freq_n, $depth_t, $freq_t);
	}

	$out;
}

# format MuTect 1.1.6/1.1.7 line
sub format_mutect_116 {
	my $line = $_[0];
	my $out;

	# split the needed columns (#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT [SAMPLE])
	my @cols = split(/\t/, $line);

	# output columns
	my $chr = $cols[0];
	my $pos = $cols[1];
	my $ref = $cols[3];
	my $alt = $cols[4];
	my $sample_t = $cols[5];
	my $sample_n = $cols[6];
	my $t_lod_fstar = sprintf("%.1f", $cols[18]);
	my $t_q20_count = $cols[24];
	my $t_ref_count = $cols[25];
	my $t_alt_count = $cols[26];
	my $tumor_f = sprintf("%.3f", $cols[21]);
	my $n_q20_count = $cols[36];
	my $n_ref_count = $cols[37];
	my $n_alt_count = $cols[38];
	my $normal_f = sprintf("%.3f", $cols[35]);

	# consider replacing raw depth with q20 depth (q20 depth and frequencies were not available in 1.1.4)

	my $depth_t = $t_alt_count + $t_ref_count;
	#my $freq_t = 0;
	#if ($depth_t > 0) {
	#	$freq_t = sprintf("%.3f", ( $t_alt_count / $depth_t ));
	#}

	my $depth_n = $n_alt_count + $n_ref_count;
	#my $freq_n = 0;
	#if ($depth_n > 0) {
	#	$freq_n = sprintf("%.3f", ( $n_alt_count / $depth_n ));
	#}

	my $judgement = $cols[50];

	if ($judgement eq "KEEP") {
		#$out = join("\t", "${chr}:${pos}:${ref}:${alt}", $sample_n, $sample_t, $chr, $pos, $t_lod_fstar, $depth_n, $freq_n, $depth_t, $freq_t);
		$out = join("\t", "${chr}:${pos}:${ref}:${alt}", $sample_n, $sample_t, $chr, $pos, $t_lod_fstar, $depth_n, $normal_f, $depth_t, $tumor_f);
	}

	$out;
}

# format MuTect2 line
sub format_mutect2 {
	my $line = $_[0];
	my $sample = $_[1];
	my $out;

	my ($sample_t, $sample_n) = split(':', $sample);

	# split the columns (#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT TUMOR NORMAL)
	my @cols = split(/\t/, $line);

	# T/N sample fields (GT:AD:AF:ALT_F1R2:ALT_F2R1:FOXOG:PGT:PID:QSS:REF_F1R2:REF_F2R1)
	my @sample_t_fields = split(':', $cols[9]);
	my @sample_n_fields = split(':', $cols[10]);

	# allelic depth field (Allelic depths for the ref and alt alleles in the order listed)
	# not using AF for now because number format is scientific for low fractions
	my @t_ad_cols = split(',', $sample_t_fields[1]);
	my @n_ad_cols = split(',', $sample_n_fields[1]);

	# output columns
	my $chr = $cols[0];
	my $pos = $cols[1];
	my $ref = $cols[3];
	my $alt = $cols[4];
	my $qual = $cols[7];
	$qual =~ s/.*TLOD=([0-9.]+).*/$1/;
	my $t_depth = $t_ad_cols[0] + $t_ad_cols[1];
	my $n_depth = $n_ad_cols[0] + $n_ad_cols[1];
	my $t_freq = $t_ad_cols[1] / $t_depth;
	my $n_freq = $n_ad_cols[1] / $n_depth;

	# report if
	# T frequency is more than 3%
	# N frequency is less than 5%
	# at least 5 variant call supporting reads
	# T frequency is sufficiently higher than N frequency
	# "we recommend applying post-processing filters, e.g. by hard-filtering calls with low minor allele frequencies"
	if ( ($t_freq > 0.03) && ($n_freq < 0.05) && ($t_ad_cols[1] >= 5) && ($t_freq > $n_freq * 5) ) {
		($pos, $ref, $alt) = adjust_indels($pos, $ref, $alt);
		$t_freq = sprintf("%.3f", $t_freq);
		$n_freq = sprintf("%.3f", $n_freq);
		my $mut_id = "${chr}:${pos}:${ref}:${alt}";
		$out = join("\t", $mut_id, $sample_t, $sample_n, $chr, $pos, $qual, $t_depth, $t_freq, $n_depth, $n_freq);
	}

	$out;
}

# format Strelka line
sub format_strelka {
	my $line = $_[0];
	my $sample = $_[1];
	my $out;

	my ($sample_t, $sample_n) = split(':', $sample);

	# split the columns (#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT NORMAL TUMOR)
	my @cols = split(/\t/, $line);

	# T/N sample fields (DP:FDP:SDP:SUBDP:AU:CU:GU:TU)
	my @sample_t_fields = split(':', $cols[10]);
	my @sample_n_fields = split(':', $cols[9]);

	# output columns
	my $chr = $cols[0];
	my $pos = $cols[1];
	my $ref = $cols[3];
	my $alt = $cols[4];
	my $qual = $cols[7];
	my $t_depth = 0;
	my $n_depth = 0;
	my $t_freq = 1;
	my $n_freq = 1;
	my $t_ref_count = 0;
	my $t_alt_count = 999;
	my $n_ref_count = 999;
	my $n_alt_count = 0;

	# SNV lines
	if ($line =~ /AU:CU:GU:TU/) {
		# allele fields (AU:CU:GU:TU)
		my %t_alleles = (
			A => (split(',', $sample_t_fields[4]))[0],
			C => (split(',', $sample_t_fields[5]))[0],
			G => (split(',', $sample_t_fields[6]))[0],
			T => (split(',', $sample_t_fields[7]))[0]
		);
		my %n_alleles = (
			A => (split(',', $sample_n_fields[4]))[0],
			C => (split(',', $sample_n_fields[5]))[0],
			G => (split(',', $sample_n_fields[6]))[0],
			T => (split(',', $sample_n_fields[7]))[0]
		);

		# output columns
		$qual =~ s/.*;QSS=(.+?);.*/$1/;
		$t_alt_count = $t_alleles{$alt};
		$n_alt_count = $n_alleles{$alt};
		$t_depth = $sample_t_fields[0];
		$n_depth = $sample_n_fields[0];
		$t_freq = $t_alt_count / $t_depth;
		$n_freq = $n_alt_count / $n_depth;
	}

	# indel lines
	if ($line =~ /TAR:TIR:TOR/) {
		# from https://github.com/Illumina/strelka/issues/3
		# somatic indel allele frequency is: alt_t1count / (ref_t1count + alt_t1count)
		# ref_t1count = 1st value of FORMAT column value TAR
		# alt_t1count = 1st value of FORMAT column value TIR

		# extract ref/alt allele counts from TAR/TIR fields
		$t_ref_count = (split(',', $sample_t_fields[2]))[0];
		$t_alt_count = (split(',', $sample_t_fields[3]))[0];
		$n_ref_count = (split(',', $sample_n_fields[2]))[0];
		$n_alt_count = (split(',', $sample_n_fields[3]))[0];

		# output columns
		$qual =~ s/.*;QSI=(.+?);.*/$1/;
		$t_depth = $t_ref_count + $t_alt_count;
		$n_depth = $n_ref_count + $n_alt_count;
		$t_freq = $t_alt_count / $t_depth;
		$n_freq = $n_alt_count / $n_depth;
	}

	# report if
	# T frequency is more than 3%
	# at least 5 variant call supporting reads
	if ( ($t_freq > 0.03) && ($t_alt_count >= 5) ) {
		($pos, $ref, $alt) = adjust_indels($pos, $ref, $alt);
		$t_freq = sprintf("%.3f", $t_freq);
		$n_freq = sprintf("%.3f", $n_freq);
		my $mut_id = "${chr}:${pos}:${ref}:${alt}";
		$out = join("\t", $mut_id, $sample_t, $sample_n, $chr, $pos, $qual, $t_depth, $t_freq, $n_depth, $n_freq);
	}

	$out;
}

# adjust indels to match annovar output
sub adjust_indels {
	my $pos = $_[0];
	my $ref = $_[1];
	my $alt = $_[2];

	# deletion (pos has to be incremented to match annovar)
	if (length($ref) > length($alt)) {
		$pos = $pos + 1;
		$ref = substr($ref, 1);
		$alt = "-";
	}

	# insertion
	if (length($alt) > length($ref)) {
		$ref = "-";
		$alt = substr($alt, 1);
	}

	return ($pos, $ref, $alt);
}



# end

File: scripts/vcf-table.R
#!/usr/bin/env Rscript


##
## Parse a VCF file and generate a table compatible with ANNOVAR output.
##
## usage: Rscript --vanilla vcf-table.R sample_name in.vcf out.txt
##


# increase output width
options(width = 120)
# print warnings as they occur
options(warn = 1)

# get scripts directory (directory of this file) and load relevant functions
args_all = commandArgs(trailingOnly = FALSE)
scripts_dir = normalizePath(dirname(sub("^--file=", "", args_all[grep("^--file=", args_all)])))
source(paste0(scripts_dir, "/load-install-packages.R"))

# relevant arguments
args = commandArgs(trailingOnly = TRUE)
sample_name = args[1]
in_vcf = args[2]
out_txt = args[3]

# check that inputs are valid
if (length(args) != 3) stop("usage: Rscript --vanilla vcf-table.R sample_name in.vcf out.txt")
if (!file.exists(in_vcf)) stop("file does not exist: ", in_vcf)

# load relevant packages
load_install_packages("dplyr")
load_install_packages("readr")
load_install_packages("tidyr")
load_install_packages("stringr")
load_install_packages("glue")
load_install_packages("vcfR")

# GATK HaplotypeCaller
parse_gatkhc = function(vcfr_obj, sample_name) {

  ##INFO DP Approximate read depth; some reads may have been filtered
  ##FORMAT AD Allelic depths for the ref and alt alleles in the order listed
  ##FORMAT DP Approximate read depth (reads with MQ=255 or with bad mates are filtered)
  ##FORMAT GQ Genotype Quality
  ##FORMAT GT Genotype

  # convert vcfR object to a tibble
  muts_tbl = vcfR2tidy(vcfr_obj,
                       info_fields = c("DP"),
                       format_fields = c("AD", "DP"),
                       single_frame = TRUE, verbose = FALSE)
  muts_tbl = muts_tbl$dat

  # extract relevant metrics
  muts_tbl =
    muts_tbl %>%
    mutate(mut_id = glue("{CHROM}:{POS}:{REF}:{ALT}")) %>%
    separate(gt_AD, into = c("ref_counts", "alt_counts"), sep = ",", convert = TRUE, extra = "drop") %>%
    mutate(AF = alt_counts / (ref_counts + alt_counts)) %>%
    mutate(
      QUAL = round(as.numeric(QUAL), 1),
      DEPTH = gt_DP,
      FREQ = round(as.numeric(AF), 3)
    )

  # adjust indels to be in ANNOVAR format
  muts_tbl = muts_tbl %>% adjust_indels()

  # manual filtering
  muts_tbl %>% filter(DEPTH >= 10 & alt_counts >= 5 & FREQ > 0.01)

}

# LoFreq (with FORMAT and SAMPLE fields added using lofreq2_add_fake_gt.py)
parse_lofreq = function(vcfr_obj, sample_name) {

  # convert vcfR object to a tibble
  muts_tbl = vcfR2tidy(vcfr_obj,
                       info_fields = c("DP", "AF", "DP4"),
                       format_fields = c("GT"),
                       single_frame = TRUE, verbose = FALSE)
  muts_tbl = muts_tbl$dat

  # extract relevant metrics
  muts_tbl =
    muts_tbl %>%
    mutate(mut_id = glue("{CHROM}:{POS}:{REF}:{ALT}")) %>%
    separate(DP4,
             into = c("ref_fwd_counts", "ref_rev_counts", "alt_fwd_counts", "alt_rev_counts"),
             sep = ",", convert = TRUE, extra = "drop") %>%
    mutate(alt_counts = alt_fwd_counts + alt_rev_counts) %>%
    mutate(
      QUAL = round(as.numeric(QUAL), 1),
      DEPTH = DP,
      FREQ = round(as.numeric(AF), 3)
    )

  # adjust indels to be in ANNOVAR format
  muts_tbl = muts_tbl %>% adjust_indels()

  # manual filtering
  muts_tbl %>% filter(DEPTH >= 10 & alt_counts >= 5 & FREQ > 0.01)

}

# Mutect 2.1 (GATK 4.0) and Mutect 2.2 (GATK 4.1)
parse_mutect2 = function(vcfr_obj, sample_T, sample_N) {

  # confirm mutect version (header changed from "Mutect Version" to "MutectVersion" in GATK 4.0.9.0)
  if (!any(str_detect(vcfr_obj@meta, "##MutectVersion=2."))) {
    stop("version mismatch: expecting Mutect 2.X")
  }

  # confirm sample names
  if (!any(str_detect(vcfr_obj@meta, glue("##tumor_sample={sample_T}")))) {
    stop("sample mismatch")
  }
  if (!any(str_detect(vcfr_obj@meta, glue("##normal_sample={sample_N}")))) {
    stop("sample mismatch")
  }

  # if a read is considered uninformative, it is counted towards the DP, but not the AD
  # an uninformative read is not reported in the AD, it is still used in calculations for genotyping
  # if uninformative reads are the only reads, we report the potential variant allele, but keep the AD values 0
  # AD is the number of reads that more likely than not support an allele
  # if you have 10 reads, each with 0.6 probability of having a certain alt allele, you get an AD of 10, whereas you get essentially 0.6 x 10 = 6 reads for the purpose of AF

  ##INFO DP Approximate read depth; some reads may have been filtered
  ##INFO TLOD Tumor LOD score
  ##FORMAT AD Allelic depths for the ref and alt alleles in the order listed
  ##FORMAT AF Allele fractions of alternate alleles in the tumor
  ##FORMAT DP Approximate read depth (reads with MQ=255 or with bad mates are filtered

  # convert vcfR object to a tibble
  muts_tbl = vcfR2tidy(vcfr_obj, single_frame = TRUE, verbose = FALSE,
                       info_fields = c("DP", "TLOD"),
                       format_fields = c("AD", "AF", "DP"))
  muts_tbl = muts_tbl$dat

  # extract relevant metrics
  muts_tbl =
    muts_tbl %>%
    # unique mutation identifier (for joining T and N)
    mutate(mut_id = glue("{CHROM}:{POS}:{REF}:{ALT}")) %>%
    filter(FILTER == "PASS") %>%
    separate(gt_AD, into = c("ref_counts", "alt_counts"), sep = ",", convert = TRUE, extra = "drop") %>%
    # manual AF calculation for comparison
    mutate(myAF = alt_counts / (ref_counts + alt_counts)) %>%
    mutate(
      QUAL = round(as.numeric(TLOD), 1),
      T_DEPTH = ref_counts + alt_counts,
      T_FREQ = round(as.numeric(gt_AF), 3)
    )

  # extract samples T and N to put side by side ("wide" format)
  snvs_n_tbl =
    muts_tbl %>%
    filter(Indiv == sample_N) %>%
    rename(N_DEPTH = T_DEPTH, N_FREQ = T_FREQ) %>%
    select(mut_id, N_DEPTH, N_FREQ)
  muts_tbl =
    muts_tbl %>%
    filter(Indiv == sample_T) %>%
    inner_join(snvs_n_tbl, by = "mut_id")

  # adjust indels to be in ANNOVAR format
  muts_tbl = muts_tbl %>% adjust_indels()

  # manual filtering
  muts_tbl %>% filter(T_DEPTH >= 10 & N_DEPTH >= 10 & alt_counts >= 5 & T_FREQ > 0.03 & T_FREQ > (N_FREQ * 5))

}

# Strelka 2
parse_strelka2 = function(vcfr_obj, sample_T, sample_N) {

  # confirm strelka version
  if (!any(str_detect(vcfr_obj@meta, "##source_version=2"))) {
    stop("version mismatch: expecting Strelka 2")
  }

  ##INFO QSS Quality score for any somatic snv
  ##INFO SOMATIC Somatic mutation">
  ##INFO QSI Quality score for the ALT haplotype to be present at a significantly different freq in the T and N
  ##FORMAT AU Number of 'A' alleles used in tiers 1,2
  ##FORMAT CU Number of 'C' alleles used in tiers 1,2
  ##FORMAT GU Number of 'G' alleles used in tiers 1,2
  ##FORMAT TU Number of 'T' alleles used in tiers 1,2
  ##FORMAT TAR Reads strongly supporting alternate allele for tiers 1,2
  ##FORMAT TIR Reads strongly supporting indel allele for tiers 1,2

  # convert vcfR object to a tibble
  muts_tbl = vcfR2tidy(vcfr_obj,
                       info_fields = c("SOMATIC", "QSS", "QSI"),
                       format_fields = c("DP", "AU", "CU", "GU", "TU", "TAR", "TIR"),
                       single_frame = TRUE, verbose = FALSE)
  muts_tbl = muts_tbl$dat
  colnames(muts_tbl)

  # extract relevant metrics
  muts_tbl =
    muts_tbl %>%
    mutate(mut_id = glue("{CHROM}:{POS}:{REF}:{ALT}")) %>%
    filter(FILTER == "PASS") %>%
    # extract tier1 counts for each nucleotide
    separate(gt_AU, into = "A_counts", sep = ",", convert = TRUE, extra = "drop") %>%
    separate(gt_CU, into = "C_counts", sep = ",", convert = TRUE, extra = "drop") %>%
    separate(gt_GU, into = "G_counts", sep = ",", convert = TRUE, extra = "drop") %>%
    separate(gt_TU, into = "T_counts", sep = ",", convert = TRUE, extra = "drop") %>%
    separate(gt_TAR, into = "indel_ref_counts", sep = ",", convert = TRUE, extra = "drop") %>%
    separate(gt_TIR, into = "indel_alt_counts", sep = ",", convert = TRUE, extra = "drop") %>%
    # set ref/alt counts
    mutate(
      ref_counts = case_when(
        is.na(QSS) ~ indel_ref_counts,
        REF == "A" ~ A_counts,
        REF == "C" ~ C_counts,
        REF == "G" ~ G_counts,
        REF == "T" ~ T_counts
      )
    ) %>%
    mutate(
      alt_counts = case_when(
        is.na(QSS) ~ indel_alt_counts,
        ALT == "A" ~ A_counts,
        ALT == "C" ~ C_counts,
        ALT == "G" ~ G_counts,
        ALT == "T" ~ T_counts
      )
    ) %>%
    # extract quality
    mutate(
      QUAL = case_when(
        is.na(QSI) ~ QSS,
        is.na(QSS) ~ QSI
      )
    ) %>%
    mutate(T_DEPTH = gt_DP) %>%
    mutate(T_FREQ = round(alt_counts / (ref_counts + alt_counts), 3))

  # extract samples T and N to put side by side ("wide" format)
  snvs_n_tbl =
    muts_tbl %>%
    filter(Indiv == "NORMAL") %>%
    rename(N_DEPTH = T_DEPTH, N_FREQ = T_FREQ) %>%
    select(mut_id, N_DEPTH, N_FREQ)
  muts_tbl =
    muts_tbl %>%
    filter(Indiv == "TUMOR") %>%
    inner_join(snvs_n_tbl, by = "mut_id")

  # adjust indels to be in ANNOVAR format
  muts_tbl = muts_tbl %>% adjust_indels()

  # manual filtering
  muts_tbl %>% filter(T_DEPTH >= 10 & N_DEPTH >= 10 & alt_counts >= 5 & T_FREQ > 0.03 & T_FREQ > (N_FREQ * 5))

}

# adjust indels to match ANNOVAR output
adjust_indels = function(x) {
  x %>%
    # set mutation type
    mutate(
      mut_type = case_when(
        str_length(REF) > str_length(ALT) ~ "del",
        str_length(ALT) > str_length(REF) ~ "ins",
        TRUE ~ "pt"
      )
    ) %>%
    # deletion (pos has to be incremented)
    mutate(
      REF = if_else(mut_type == "del", str_sub(REF, 2), REF),
      ALT = if_else(mut_type == "del", "-", ALT),
      POS = if_else(mut_type == "del", as.integer(POS + 1), POS)
    ) %>%
    # insertion
    mutate(
      REF = if_else(mut_type == "ins", "-", REF),
      ALT = if_else(mut_type == "ins", str_sub(ALT, 2), ALT)
    )
}

# split sample name for somatic variants
if (str_detect(sample_name, ":")) {
  sample_T = str_split_fixed(sample_name, ":", 2)[1]
  sample_N = str_split_fixed(sample_name, ":", 2)[2]
}

# import VCF as a vcfR object
muts_vcfr = read.vcfR(in_vcf, verbose = FALSE)

# check if there are any variants
if (nrow(muts_vcfr@fix) == 0) stop("no variants in imported VCF")

# determine variant caller based on VCF contents and parse accordingly
if (any(str_detect(muts_vcfr@meta, "##GATKCommandLine.HaplotypeCaller"))) {

  # GATK HaplotypeCaller
  message("parsing GATK HaplotypeCaller VCF")
  caller_type = "germline"
  vcf_tbl = parse_gatkhc(vcfr_obj = muts_vcfr, sample_name = sample_name)

} else if (any(str_detect(muts_vcfr@meta, "##source=lofreq"))) {

  # LoFreq
  message("parsing LoFreq VCF")
  caller_type = "germline"
  vcf_tbl = parse_lofreq(vcfr_obj = muts_vcfr, sample_name = sample_name)

} else if (any(str_detect(muts_vcfr@meta, "##source=Mutect2"))) {

  # Mutect 2.1 (GATK 4.0) or Mutect 2.2 (GATK 4.1)
  message("parsing Mutect 2.X VCF")
  caller_type = "somatic"
  vcf_tbl = parse_mutect2(vcfr_obj = muts_vcfr, sample_T = sample_T, sample_N = sample_N)

} else if (any(str_detect(muts_vcfr@meta, "##source=strelka"))) {

  # Strelka 2
  message("parsing Strelka 2 VCF")
  caller_type = "somatic"
  vcf_tbl = parse_strelka2(vcfr_obj = muts_vcfr, sample_T = sample_T, sample_N = sample_N)

} else {

  stop("unknown variant caller")

}

# check if table is empty
if (nrow(vcf_tbl) == 0) stop("output table is empty after parsing")

# update locale for sorting (to match the rest of the pipeline)
Sys.setlocale(category = "LC_ALL", locale = "C")

# create and sort by #MUT
vcf_tbl =
  vcf_tbl %>%
  mutate(`#MUT` = glue("{CHROM}:{POS}:{REF}:{ALT}")) %>%
  rename(CHR = CHROM) %>%
  arrange(`#MUT`)

# determine output table columns
if (caller_type == "germline") {
  vcf_tbl = vcf_tbl %>% mutate(SAMPLE = sample_name)
  out_cols = c("#MUT", "SAMPLE", "CHR", "POS", "QUAL", "DEPTH", "FREQ")
}
if (caller_type == "somatic") {
  vcf_tbl = vcf_tbl %>% mutate(SAMPLE_T = sample_T, SAMPLE_N = sample_N)
  out_cols = c("#MUT", "SAMPLE_T", "SAMPLE_N", "CHR", "POS", "QUAL", "T_DEPTH", "T_FREQ", "N_DEPTH", "N_FREQ")
}

# keep only relevant columns
clean_vcf_tbl = vcf_tbl[, out_cols]

# export
write_tsv(clean_vcf_tbl, path = out_txt)



# end
